pkg/storage/etcd3/store.go

```diff
@@ -346,19 +358,35 @@ func (s *store) getState(getResp *clientv3.GetResponse, key string, v reflect.Va
 	return state, nil
 }
 
-func (s *store) updateState(st *objState, userUpdate storage.UpdateFunc) (runtime.Object, error) {
-	ret, _, err := userUpdate(st.obj, *st.meta)
+func (s *store) updateState(st *objState, userUpdate storage.UpdateFunc) (runtime.Object, uint64, error) {
+	ret, ttlPtr, err := userUpdate(st.obj, *st.meta)
+
 	version, err := s.versioner.ObjectResourceVersion(ret)
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
 	if version != 0 {
 		// We cannot store object with resourceVersion in etcd. We need to reset it.
 		if err := s.versioner.UpdateObject(ret, nil, 0); err != nil {
-			return nil, fmt.Errorf("UpdateObject failed: %v", err)
+			return nil, 0, fmt.Errorf("UpdateObject failed: %v", err)
 		}
 	}
-	return ret, nil
+	var ttl uint64
+	if ttlPtr != nil {
+		ttl = *ttlPtr
+	}
+	return ret, ttl, nil
+}
+
+func (s *store) createLeaseOpt(ctx context.Context, ttl int64) (clientv3.OpOption, error) {
```

Comment:
one lease per ttl key is expensive. i would suggest to great one lease for every 10 minutes window. at lease, we should add a todo.